polkadot/caip2.md

```diff
@@ -0,0 +1,109 @@
+---
+namespace-identifier: polkadot-caip2
+title: Polkadot Namespace - Chains
+author: ["Pedro Gomes (@pedrouid)", "Joshua Mir (@joshua-mir)", "Shawn Tabrizi (@shawntabrizi)", "Juan Caballero (@bumblefudge)"]
+discussions-to: https://github.com/ChainAgnostic/CAIPs/issues/13
+status: Draft
+type: Standard
+created: 2020-04-01
+updated: 2022-03-27
+requires: CAIP-2
+replaces: CAIP-13
+---
+
+# CAIP-2
+
+*For context, see the [CAIP-2][] specification.*
+
+## Rationale
+
+The namespace is called "polkadot" to refer to Polkadot-like "parachains,"
+parallel and independent Layer 1s coordinated by the Polkadot relay chains.
+
+The rationale behind the use of block hash from the genesis block stems from its
+usage in the Polkadot architecture in network and consensus.
+
+## Syntax
+
+The definition for this namespace will use the `genesis-hash` as an indentifier
+for different Polkadot chains. The format is a 32 character prefix of the block
+hash (lower case SHA256 hex).
+
+As CAIP-2 references for this namespace are 32-byte SHA256 hashes in lowercase
+hex, they can be validated with the simple regular expression: `[0-9a-f]{32}`
+
+### Resolution Method
+
+To resolve a blockchain reference for the Polkadot namespace, make a JSON-RPC
+request to the blockchain node with method `chain_getBlockHash`, for example:
+
+```jsonc
+// Request
+{
+  "id": 1,
+  "jsonrpc": "2.0",
+  "method": "chain_getBlockHash",
+  "params": [0]
+}
+
+// Response
+{
+  "id": 1,
+  "jsonrpc": "2.0",
+  "result": "0x91b171bb158e2d3848fa23a9f1c25182fb8e20313b2c1eb49219da7a70ce90c3"
+}
+```
+The response will return as a value for the result a hash for the block with
+height 0 that should be sliced to its first 16 bytes (32 characters for base 16)
+to be CAIP-2/CAIP-10 compatible.
+
+### Backwards Compatibility
+
+Not applicable
+
+## Test Cases
+
+This is a list of manually composed examples
+
+```
+# Kusama
+polkadot:b0a8d493285c2df73290dfb7e61f870f
+
+# Edgeware
+polkadot:742a2ca70c2fda6cee4f8df98d64c4c6
+
+# Kulupu
+polkadot:37e1f8125397a98630013a4dff89b54c
+```
```

Comment:
"The namespace is called "polkadot" to refer to Polkadot-like "parachains,""

None of these are parachains :). Ideally namespaces would anchor at consensus systems, so e.g.:

```
polkadot:<statemint-genesis-hash>
kusama:<statemine-genesis-hash>
```

Or, if you wanted to express that Kusama and Polkadot are connected, perhaps add another "relational" namespace in the path, like:

```
polkadot:canary:<kusama-genesis-hash>
polkadot:para:<statemint-genesis-hash>
polkadot:solo:<kulupu-genesis-hash>
```

The problem that will come up is migrations. For example, Centrifuge and Integritee (maybe some others, these two off the top of my head) were solo chains that became parachains. So if these identifiers are meant to last forever, there could be issues, but I'm really not that familiar with CAIP.

Or, if you want to avoid migration issues entirely, just change the docs from "The namespace is called "polkadot" to refer to Polkadot-like "parachains,"" to something like, "The namespace is called "polkadot" to refer to blockchains whose state transition functions can execute on the Polkadot host API, for example parachains or Kusama."

Reply:
In the latter example, would all chains then still use the `polkadot` namespace? I saw a comment about the `MultiLocation` struct that mentions forks also for (para)chains in the Polkadot space. If that is the case, I think having a genesis hash as a way to identify chains is definitely not sufficient, as two forked chains would be identified by the same hash. To this point, would a combination of genesis hash and SS58 prefix be sufficient to uniquely identify all chains, including forks and migrated ones?

Reply:
> In the latter example, would all chains then still use the polkadot namespace? 

Yes, as they are chains that execute on top of the Polkadot host API (which is not just expressed in Substrate, but also in other implementations like ChainSafe's Gossamer).

> I saw a comment about the MultiLocation struct that mentions forks also for (para)chains in the Polkadot space. If that is the case, I think having a genesis hash as a way to identify chains (as is the case now) is definitely not sufficient, as two forked chains would be identified by the same hash.

How do you define a fork? A parachain should never fork, unless someone starts a new chain with a modified chainspec/runtime, but then that's just a new genesis. `MultiLocation`s in general are _relative_ location paths but don't actually know about a `/` location (although in practice it would be the governing consensus system). Since CAIP does seem to have a `/` location (the first namespace), `MultiLocation` could be mapped onto it from one chain ID to another.

> would a combination of genesis hash and SS58 prefix be sufficient to uniquely identify all chains, including forks and migrated ones?

No, SS58 should probably not even come near this conversation. It's 99% for front end, there is no guarantee that it's unique (already, Polkadot and Statemint both use a prefix of 0), nor that it will stay the same in a chain's lifetime.

Reply:
Sorry, my mistake. Forks could nevertheless still exist for standalone chains, and if they would be identified with the same `polkadot` namespace, then using the sole genesis hash would not be sufficient anymore.

> No, SS58 should probably not even come near this conversation. It's 99% for front end, there is no guarantee that it's unique (already, Polkadot and Statemint both use a prefix of 0), nor that it will stay the same in a chain's lifetime.

Yeah that's what I thought. I wanted to make it clear as there has been conversation up above about using those to identify chains.

Reply:
some time next month I'll devote a big bloc of time to overhauling all this and tag you both in the PR-- I clearly need to go back to the drawing board to map the CAIP mental model onto this VM!

Reply:
Going to merge a straw man and invite PRs to improve, will return to this discussion